source/extensions/tracers/opencensus/opencensus_tracer_impl.cc

```diff
@@ -0,0 +1,171 @@
+
+void Span::injectContext(Http::HeaderMap& request_headers) {
+  if (oc_config_->propagate_cloud_trace_context()) {
+    request_headers.addCopy(
+        Http::LowerCaseString{"x-cloud-trace-context"},
+        ::opencensus::trace::propagation::ToCloudTraceContextHeader(span_.context()));
+  }
+  if (oc_config_->propagate_trace_context()) {
+    request_headers.addCopy(Http::LowerCaseString{"traceparent"},
+                            ::opencensus::trace::propagation::ToTraceParentHeader(span_.context()));
+  }
+  if (oc_config_->propagate_grpc_trace_bin()) {
+    request_headers.addCopy(
+        Http::LowerCaseString{"grpc-trace-bin"},
+        ::opencensus::trace::propagation::ToGrpcTraceBinHeader(span_.context()));
+  }
+}
+
+Tracing::SpanPtr Span::spawnChild(const Tracing::Config& /*config*/, const std::string& name,
+                                  SystemTime /*start_time*/) {
+  span_.AddAnnotation("spawnChild");
+  return std::make_unique<Span>(oc_config_,
+                                ::opencensus::trace::Span::StartSpan(name, /*parent=*/&span_));
+}
+
+void Span::setSampled(bool sampled) { span_.AddAnnotation("setSampled", {{"sampled", sampled}}); }
+
+} // namespace
+
+Driver::Driver(const envoy::config::trace::v2::OpenCensusConfig& oc_config)
+    : oc_config_(oc_config) {
+  if (oc_config.has_trace_config()) {
+    ApplyTraceConfig(oc_config.trace_config());
+  }
+  if (oc_config.stdout_exporter_enabled()) {
+    ::opencensus::exporters::trace::StdoutExporter::Register();
+  }
+  if (oc_config.stackdriver_exporter_enabled()) {
+    ::opencensus::exporters::trace::StackdriverOptions opts;
+    opts.project_id = oc_config.stackdriver_project_id();
+    ::opencensus::exporters::trace::StackdriverExporter::Register(opts);
+  }
+  if (oc_config.zipkin_exporter_enabled()) {
+    ::opencensus::exporters::trace::ZipkinExporterOptions opts(oc_config.zipkin_url());
+    opts.service_name = oc_config.zipkin_service_name();
+    ::opencensus::exporters::trace::ZipkinExporter::Register(opts);
+  }
+}
+
+void Driver::ApplyTraceConfig(const opencensus::proto::trace::v1::TraceConfig& config) {
+  using SamplerCase = opencensus::proto::trace::v1::TraceConfig::SamplerCase;
+  constexpr double kDefaultSamplingProbability = 1e-4;
+  double probability = kDefaultSamplingProbability;
+
+  switch (config.sampler_case()) {
+  case SamplerCase::kProbabilitySampler:
+    probability = config.probability_sampler().samplingprobability();
+    break;
+  case SamplerCase::kConstantSampler:
+    probability = config.constant_sampler().decision() ? 1. : 0.;
+    break;
+  case SamplerCase::kRateLimitingSampler:
+    ENVOY_LOG(error, "RateLimitingSampler is not supported.");
+    break;
+  case SamplerCase::SAMPLER_NOT_SET:
+    break; // Keep default.
+  default:
+    ENVOY_LOG(error, "Unknown sampler type in TraceConfig.");
+  }
+
+  ::opencensus::trace::TraceConfig::SetCurrentTraceParams(::opencensus::trace::TraceParams{
```

Comment:
Is this a global/singleton setting? We generally prefer not to do this in Envoy, for testability etc.

Reply:
Fair enough; since we only have one HTTP tracer per Envoy instance, this seems safe.